rfcs/20221020-tensor-indexed-updates.md

```diff
@@ -0,0 +1,438 @@
+# Construct a set of indices to access tensor[:,:,2].
+x_shape = tf.shape(x)
+dim0 = tf.range(x_shape[0])
+dim1 = tf.range(x_shape[1])
+dim0, dim1 = tf.meshgrid(dim0, dim1, indexing='ij')
+dim0, dim1 = tf.reshape(dim0, [-1]), tf.reshape(dim1, [-1])
+dim_shape = tf.shape(dim0)
+dim2 = tf.broadcast_to(tf.constant(2, dtype=dim0.dtype), dim_shape)
+indices = tf.transpose(tf.stack([dim0, dim1, dim2]))
+
+# Create a set of values with the same number of elements in indices, since
+# tensor_scatter_* currently doesn't support broadcasting.
+updates = tf.broadcast_to(tf.constant(0, dtype=tensor.dtype),dim_shape)
+
+# Scatter updates into x.
+x = tf.tensor_scatter_nd_update(x, indices, updates)
+```
+
+**Setting negative values to zero:**
+```python
+#========================================================================
+# Numpy:
+#========================================================================
+x[x < 0] = 0
+
+#========================================================================
+# TensorFlow via tf.tensor_scatter_nd_update:
+#========================================================================
+# Extract index tensor from boolean mask.
+indices = tf.where(x < 0)
+# Build values tensor.
+dim_size = tf.shape(indices)[0:1]
+values = tf.broadcast_to(tf.constant(0, dtype=x.dtype), dim_size)
+# Scatter updates into x.
+x = tf.tensor_scatter_nd_update(x, indices, values)
+
+#========================================================================
+# TensorFlow via multiplexing tf.where:
+#========================================================================
+# NOTE: this is only equivalent to NumPy for simple cases where the "updates"
+# is a scalar.  Otherwise, the RHS to `x[mask] = RHS` needs to match the shape
+# of indices returned via `tf.where(mask)`, which is not the same shape as the
+# mask itself.
+x = tf.where(x < 0, 0, x)
+```
+
+Although it still is technically possible to perform many of the same operations
+with using TensorFlow's existing functionality, it's quite cumbersome.
+
+## Proposed Syntax
+
+Rather than invent a new API, we will rely on the successful rollout of indexed
+updates in JAX - which also has the immutable tensor restriction - via the
+[`.at[]`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html#jax.numpy.ndarray.at) syntax.
+
+
+```python
+#========================================================================
+# Proposed TensorFlow Syntax:
+#========================================================================
+# Setting a tensor slice to zero:
+x = x.at[:,:,2].set(0)
+
+# Setting negative values to zero:
+x = x.at[x < 0].set(0)
+```
+
+For `tf.Variable`, we have already had _some_ support for in-place slice updates
+via the `.assign()` method:
+```python
+# Existing tf.Variable in-place slice assignment:
+v[:, :, 2].assign(0)
+```
+For consistency with NumPy, we propose to allow direct assignment as well:
```

Comment:
This would make tf.Variable consistent with NumPy, but different from Tensors. WDYT about making tf.Variable consistent with Tensors instead? e.g.
`v.at[:, :, 2].set(0)`

Maybe this would simplify the users' mental model a bit

Reply:
```
v.at[:, :, 2].set(0)
```
is proposed here, but it returns a _copy_ of the variable, just as `Tensor` returns a copy.  It does not modify in-place.

One of the comments from the internal design review was that Rohan and Mark wanted a more numpy-like syntax when it is supported.  So for `tf.Variable`, we are actually consistent with both.

Reply:
Sorry I missed it! SG.

Reply:
Added this more explicitly just below.